Gradle fails with "Could not apply requested plugin" in some scenarios using configuration on demand

I'm seeing a bug where a Gradle plugin defined in an included build applied to multiple project fails in certain scenarios (when certain tasks are requested). 

### Expected Behavior
Given each of the projects independently builds fine, I would expect requesting several of them to build at the same time would also build fine.

### Current Behavior
The build fails with an error like the following:

```
❯ ./gradlew :projects:baz:classes :projects:foo:bar:classes --scan
Configuration on demand is an incubating feature.

> Configure project :projects:foo:bar
Hello, world! Applying to bar

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/sean_abraham/projects/personal/gradle-repros/included-build-plugin-application/projects/foo/build.gradle' line: 4

* What went wrong:
Could not apply requested plugin [id: 'com.example.buildplugin'] as it does not provide a plugin with id 'com.example.buildplugin'. This is caused by an incorrect plugin implementation. Please contact the plugin author(s).
> Plugin with id 'com.example.buildplugin' not found.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 2s
```

When both `:projects:baz` and `:projects:foo:bar` each build find (as well as `projects:foo:` which `:projects:baz` depends on). Note this both fails using the `plugins { }` block as well as legacy plugin application format (buildscript classpath dependencies + `apply plugin ...`).

See the included repro for further details

### Context
This is blocking my ability to refactor a module into two, both of which need the same build plugin defined (which is defined in an included build).

### Steps to Reproduce 
https://github.com/seanabraham/included-build-plugin-gradle-repro

### Your Environment
Build scan URL: https://scans.gradle.com/s/w4yopkrbla6lg

--- a/subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/CompositeBuildPluginDevelopmentIntegrationTest.groovy
+++ b/subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/CompositeBuildPluginDevelopmentIntegrationTest.groovy
@@ -483,6 +483,34 @@ class CompositeBuildPluginDevelopmentIntegrationTest extends AbstractCompositeBu
         failure.assertHasDescription("Plugin [id: 'org.test.plugin.pluginBuild'] was not found in any of the following sources:")
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/14552")
+    @ToBeFixedForConfigurationCache
+    def "can co-develop plugin with nested consumers using configure-on-demand"() {
+        given:
+        buildA = multiProjectBuild("cod", ["foo", "foo:bar"])
+        includeBuild pluginBuild
+
+        buildA.file("foo/build.gradle") << """
+plugins {
+    id 'java-library'
+    id 'org.test.plugin.pluginBuild'
+}
+"""
+        buildA.file('foo/bar/build.gradle') << """
+plugins {
+    id 'java-library'
+    id 'org.test.plugin.pluginBuild'
+}
+"""
+
+        when:
+        args "--configure-on-demand"
+        execute(buildA, ":foo:bar:classes", ":foo:classes")
+
+        then:
+        executed ":pluginBuild:jar", ":foo:classes", ":foo:bar:classes"
+    }
+
     def addLifecycleTasks(BuildTestFile build) {
         build.buildFile << """
             tasks.maybeCreate("assemble")

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/ConfigurationOnDemandIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/ConfigurationOnDemandIntegrationTest.groovy
@@ -512,4 +512,23 @@ allprojects {
         result.assertTasksExecuted(":a:one")
         fixture.assertProjectsConfigured(":", ":b", ":b:child", ":a")
     }
+
+    def "extra properties defined in parent project are accessible to child"() {
+        settingsFile << "include 'a', 'a:child'"
+        file('a/build.gradle') << """
+ext.foo = "Moo!!!"
+"""
+        file('a/child/build.gradle') << """
+task printExt {
+    doLast {
+        println "The Foo says " + foo
+    }
+}
+"""
+        when:
+        run(":a:child:printExt")
+
+        then:
+        outputContains("The Foo says Moo!!!")
+    }
 }

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/ConfigurationOnDemandIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/ConfigurationOnDemandIntegrationTest.groovy
@@ -72,18 +72,14 @@ class ConfigurationOnDemandIntegrationTest extends AbstractIntegrationSpec {
     }
 
     def "evaluates only project referenced in the task list"() {
-        // The util project's classloaders will be created eagerly because util:impl
-        // will be evaluated before it
-        executer.withEagerClassLoaderCreationCheckDisabled()
-
         settingsFile << "include 'api', 'impl', 'util', 'util:impl'"
         buildFile << "allprojects { task foo }"
 
         when:
-        run(":foo", ":util:impl:foo")
+        run(":util:impl:foo")
 
         then:
-        fixture.assertProjectsConfigured(":", ":util:impl")
+        fixture.assertProjectsConfigured(":", ":util", ":util:impl")
     }
 
     def "does not show configuration on demand incubating message in a regular mode"() {

--- a/subprojects/core/src/main/java/org/gradle/execution/TaskPathProjectEvaluator.java
+++ b/subprojects/core/src/main/java/org/gradle/execution/TaskPathProjectEvaluator.java
@@ -33,6 +33,11 @@ public class TaskPathProjectEvaluator implements ProjectConfigurer {
         if (cancellationToken.isCancellationRequested()) {
             throw new BuildCancelledException();
         }
+        // Need to configure intermediate parent projects for configure-on-demand
+        ProjectInternal parentProject = project.getParent();
+        if (parentProject != null) {
+            configure(parentProject);
+        }
         project.evaluate();
     }
 

--- a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/build/BuildTestFixture.groovy
+++ b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/build/BuildTestFixture.groovy
@@ -90,7 +90,7 @@ class BuildTestFixture {
         rootMulti.with(cl)
         rootMulti.file("src/main/${language.name}/Dummy.${language.name}") << "public class Dummy {}"
         subprojects.each {
-            rootMulti.file(it, "src/main/${language.name}/Dummy.${language.name}") << "public class Dummy {}"
+            rootMulti.file(it.replace(':' as char, File.separatorChar), "src/main/${language.name}/Dummy.${language.name}") << "public class Dummy {}"
         }
         return rootMulti
     }

